bluesky_widgets/qt/search_results.py

```diff
@@ -34,20 +40,22 @@ class DataLoader(QThread):
       viewer from the data cache.
     """
 
-    def __init__(self, queue, get_data, data, data_changed, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self._queue = queue
+    def __init__(self, get_data, data, *args, **kwargs):
+        super().__init__(*args, SignalsClass=DataLoaderSignals, **kwargs)
         self._get_data = get_data
         self._data = data
-        self._data_changed = data_changed
+        self._queue = queue.Queue()
 
-    def run(self):
+    def request(self, work):
```

Comment:
Can def request be turned into a slot, that way you can emit from _fetchMore a signal that is received by the request slot? Otherwise you are calling into request in DataLoader which touches self._queue from another thread

Reply:
I realized that you are running a QRunnable object and QThread anymore, which complicates this idea slightly. The recommendation from Qt in this case is that you are on your own to guard your data using Mutexes.

Reply:
I don't fully understand how using the QThreadPool instead of QThread complicates or changes things. Can you explain more? In any case, I'm sold that using signals/slots would be better and is worth doing anyway.